lib/bson/dbref.rb

```diff
@@ -122,6 +122,26 @@ def from_bson(buffer, **options)
         decoded
       end
     end
+
+    private
+
+    # Reorder the fields of the given Hash to have $ref first, $id second,
+    # and $db third. The rest of the fields in the hash can come in any
+    # order after that.
+    #
+    # @param [ Hash ] hash The input hash. Must be a valid dbref.
+    #
+    # @return [ Hash ] The hash with it's fields reordered.
+    def reorder_fields(hash)
+      reordered = {}
+      reordered['$ref'] = hash.delete('$ref')
```

Comment:
This mutates the input which is not good, easiest way to fix this is to create a `BSON::Document` out of the `hash` which would also provide indifferent access for validation in the constructor.

Reply:
done